File: ga/4omini_result_2024_7/result/1968G1_mini/38_1968G1.c
use std::io::{self, BufRead};
use rand::Rng;  // Import the RNG functionality from the rand crate

const N: usize = 200005;
const MOD: i64 = 1_000_000_007;

fn main() {
    let stdin = io::stdin();
    let reader = stdin.lock();
    let mut lines = reader.lines();

    let t: usize = lines.next().unwrap().unwrap().parse().unwrap();
    
    for _ in 0..t {
        // Read n, l, r, and s
        let input: Vec<_> = lines.next().unwrap().unwrap().split_whitespace().collect();
        let n: usize = input[0].parse().unwrap();
        let l: usize = input[1].parse().unwrap();
        let r: usize = input[2].parse().unwrap();
        let s: Vec<char> = format!("0{}", input[3]).chars().collect(); // Add extra index for easier handling (1-indexed)

        let base = rand::thread_rng().gen_range(200..MOD as u64) as i64;

        let mut base_pow = vec![1; n + 1];
        let mut hash = vec![0; n + 1];
        let mut bp: i64 = 1;
        
        for i in 1..=n {
            bp = (bp * base) % MOD;
            base_pow[i] = bp;
            hash[i] = (hash[i - 1] + (s[i] as i64 * base_pow[i]) % MOD) % MOD;
        }

        let (mut lower, mut upper) = (0, n + 1);
        let mut yy: i64;

        while upper - lower > 1 {
            let mid = lower + (upper - lower) / 2;
            yy = check(mid, n, &hash, &base_pow);
            if yy >= l as i64 {
                lower = mid;
            } else {
                upper = mid;
            }
        }
        print!("{} ", lower);

        for i in (l + 1)..=r {
            if i > yy {
                upper = lower + 1;
                lower = 0;

                while upper - lower > 1 {
                    let mid = lower + (upper - lower) / 2;
                    yy = check(mid, n, &hash, &base_pow);
                    if yy >= i as i64 {
                        lower = mid;
                    } else {
                        upper = mid;
                    }
                }
            }
            print!("{} ", lower);
        }
        println!();
    }
}

fn check(k: usize, n: usize, hash: &Vec<i64>, base_pow: &Vec<i64>) -> i64 {
    let mut vo = 0;
    let mut i = 1;
    
    while i + k - 1 <= n {
        let h1 = (hash[i + k - 1] + MOD - hash[i - 1]) % MOD;
        let h2 = (hash[k] * base_pow[i - 1]) % MOD;
        if h1 == h2 {
            vo += 1;
            i += k;  // Non-overlapping
        } else {
            i += 1;  // Overlapping
        }
    }
    vo
}

File: ga/glm4_result_2024_6/result/1968G1_cpp_glm/38_1968G1.rs
use std::io::{self, BufRead};

fn z_function(s: &str) -> Vec<usize> {
    let n = s.len();
    let mut z = vec![0; n];
    let mut l = 0;
    let mut r = 0;
    for i in 1..n {
        if i <= r && z[i - l] < r - i + 1 {
            z[i] = z[i - l];
        } else {
            z[i] = std::cmp::max(0, r - i + 1);
            while i + z[i] < n && s.chars().nth(z[i]) == s.chars().nth(i + z[i]) {
                z[i] += 1;
            }
        }
        if i + z[i] - 1 > r {
            l = i;
            r = i + z[i] - 1;
        }
    }
    z[0] = n;
    z
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();
    let t: usize = lines.next().unwrap().unwrap().parse().unwrap();
    for _ in 0..t {
        let mut line = lines.next().unwrap().unwrap();
        let mut parts = line.split_whitespace();
        let n: usize = parts.next().unwrap().parse().unwrap();
        let l: usize = parts.next().unwrap().parse().unwrap();
        let r: usize = parts.next().unwrap().parse().unwrap();
        let s: String = parts.next().unwrap().to_string();

        let mut ls = 0;
        let mut rs = n + 1;
        let z = z_function(&s);
        while ls + 1 < rs {
            let mid = (ls + rs) / 2;
            if (0..n/mid).any(|i| z[i * mid] >= mid) {
                ls = mid;
            } else {
                rs = mid;
            }
        }
        println!("{}", ls);
    }
}